File: leet-code/973. K Closest Points to Origin/KClosestPointsToOrigin973.java
import java.util.Comparator;
import java.util.PriorityQueue;

public class KClosestPointsToOrigin973 {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> pq = new PriorityQueue<>(Comparator.comparingInt(p -> p.distance));
        for (int[] point : points) {
            pq.add(new Point(point));
        }

        int[][] answer = new int[k][2];
        for (int i = 0; i < k; i++) {
            answer[i] = pq.poll().idx;
        }
        return answer;
    }

    private static class Point {
        int[] idx;
        int distance;

        Point(int[] idx) {
            this.idx = idx;
            distance = (idx[0] * idx[0]) + (idx[1] * idx[1]);
        }
    }
}

File: leet-code/973. K Closest Points to Origin/KClosestPointsToOrigin973.kt
import java.util.PriorityQueue

class KClosestPointsToOrigin973_2 {
    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val pq: PriorityQueue<Point> = PriorityQueue<Point>(Comparator.comparingInt{ it.distance })
        for (point in points) {
            pq.add(Point(point))
        }
        val answer: Array<IntArray> = Array(k) { intArrayOf() }
        for (i in 0 until k) {
            answer[i] = pq.poll().point
        }
        return answer
    }

    class Point(val point: IntArray) {
        val distance: Int = (point[0] * point[0]) + (point[1] * point[1])
    }
}